### ISSUE TITLE
Remove docker dependency on Kubelet startup

### ISSUE THREAD

**Issue by dchen1107** [kubernetes/kubernetes#54090]
We removed docker dependencies from kubelet /healthz a couple of releases before, so that kubelet can properly report runtime is down. But Kubelet still have a docker dependency at initial stage: 

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L150

```
	var dockerClient libdocker.Interface
	if s.ContainerRuntime == kubetypes.DockerContainerRuntime {
		dockerClient = libdocker.ConnectToDockerOrDie(s.DockerEndpoint, s.RuntimeRequestTimeout.Duration,
			s.ImagePullProgressDeadline.Duration)
	} else {
		dockerClient = nil
	}
```
We should consider removing such dependency. 

cc/ @kubernetes/sig-node-bugs

**Comment by yujuhong** [comment]
Also need to remove dockershim's dependency on docker at startup.

**Comment by resouer** [comment]
Allow me to take a deep look on this, cc @yujuhong

**Comment by luxas** [comment]
@resouer Can you fix this in v1.9 please?

**Comment by resouer** [comment]
@luxas NP, already discussed with yuju about the idea and patch is by handy.

**Comment by resouer** [comment]
Let's split this task into several steps (thanks Yuju's advice):

```
1) make kubelet (minus dockershim) not depend on docker at startup #54405
2) make dockershim not depend on docker at startup. (already have drafted design)

The first one is desirable, while the second is just for practical purpose since dockershim is part of kubelet.
```

**Comment by dims** [comment]
@resouer Looks like #54405 went in, but the step 2 is not filed? (So can we move this to 1.10)?

**Comment by resouer** [comment]
@dims Yes, 1.10 is the plan.

**Comment by jberkus** [comment]
has there been any work on this since December?  If not, please remove from the 1.10 milestone.

@dims @resouer

**Comment by yujuhong** [comment]
@resouer is this still targeted for 1.10?

**Comment by resouer** [comment]
@yujuhong Sorry for missing your message >_<

Actually, someone else claimed this issue from me a month ago but seems he did not make it. I will pick this up in next cycle.

**Comment by dims** [comment]
@resouer do we still want to do this?

**Comment by pdxkzeerkti** [comment]
I don't know whether commenting here is the right approach but I'm going to start with this comment in preference to filing a new issue.

Today, I found it necessary to modify the Docker configuration (/etc/docker/docker.json) on all my nodes. In each case, I made the change and restarted the Docker service using `systemctl`. 

Exactly one of my three nodes went **NotReady** within seconds of restarting the Docker service. Journalctl on that node showed:
```
Sep 11 16:23:26 host[1]: Stopping Kubernetes Kubelet...
Sep 11 16:23:26 host[1]: Stopped Kubernetes Kubelet.
Sep 11 16:23:36 host[1]: Dependency failed for Kubernetes Kubelet.
```
Examination showed that all Kubernetes-related processes on the node had exited.

I understand about fail-stop behavior and nodes being cattle rather than pets, but losing the node still seems like an excessive consequence of merely needing to change a service configuration. I'm using 1.13.5 on Oracle Linux 7.6. The host is a VM in the Oracle cloud (OCI).
